gap/semigroups/semieunit.gi

```diff
@@ -0,0 +1,715 @@
+#############################################################################
+##
+#W  semieunit.gi
+#Y  Copyright (C) 2017                                    Christopher Russell
+##
+##  Licensing information can be found in the README file of this package.
+##
+#############################################################################
+#
+# This code is based on the description of McAlister triple semigroups found
+# in Fundamentals of Semigroup Theory by John Howie.
+#
+#############################################################################
+# Methods for creating McAlister triple semigroups
+#############################################################################
+InstallMethod(McAlisterTripleSemigroup,
+"for a perm group, action, digraph, and digraph",
+[IsGroup, IsFunction, IsDigraph, IsDigraph],
+function(G, act, X, Y)
+  local anti_act, hom, y_edges, out_nbrs, orbs, min, fam, filt, M, x, y;
+
+  if not IsFinite(G) then
+    ErrorNoReturn("Semigroups: McAlisterTripleSemigroup: usage,\n",
+                  "the first argument must be a finite group,");
+  fi;
+
+  anti_act := function(pt, g)
+    return act(pt, g ^ -1);
+  end;
+
+  hom := ActionHomomorphism(G, DigraphVertices(X), anti_act);
+
+  if not IsSubgroup(AutomorphismGroup(X), Image(hom)) then
+    ErrorNoReturn("Semigroups: McAlisterTripleSemigroup: usage,\n",
+                  "the first argument (a group) must act by order ",
+                  "automorphisms on the third argument (a partial order ",
+                  "digraph),");
+  elif not IsPartialOrderDigraph(X) then
+    ErrorNoReturn("Semigroups: McAlisterTripleSemigroup: usage,\n",
+                  "the third argument must be a partial order digraph,");
+  fi;
+
+  # To check if the edges of y are a subset of the edges of x we need to apply
+  # the labeling of y's vertices to its edges.
+  y_edges := List(DigraphEdges(Y), edge -> [DigraphVertexLabels(Y)[edge[1]],
+                                            DigraphVertexLabels(Y)[edge[2]]]);
+
+  if not (IsSubset(DigraphVertexLabels(X), DigraphVertexLabels(Y)) and
+          IsSubset(DigraphEdges(X), y_edges)) then
+    ErrorNoReturn("Semigroups: McAlisterTripleSemigroup: usage,\n",
+                  "the fourth argument must be a subdigraph of the third ",
+                  "argument,");
+  elif not IsJoinSemilatticeDigraph(Y) then
+    ErrorNoReturn("Semigroups: McAlisterTripleSemigroup: usage,\n",
+                  "the fourth argument must be a join semilattice digraph,");
+  fi;
+
+  # Check condition M2 (check that Y is an order ideal of X.)
+  # TODO: implement IsOrderIdeal for a subset of a partial order digraph.
+  out_nbrs := OutNeighbors(X);
+  for x in DigraphVertexLabels(X) do
+    if not x in DigraphVertexLabels(Y) then
+      for y in DigraphVertexLabels(Y) do
+        if Intersection(out_nbrs[x], out_nbrs[y]) = out_nbrs[x] then
```

Comment:
There's also another vertex number/label of `X` conundrum here: `x` is one of the vertex labels of `X`, *not* necessarily the actualy vertex number. However, `out_nbrs[x]` gives the out-neighbours (as vertex numbers, not labels) of the vertex whose number (not label) is `x`. Do you see the problem?

i.e `x` might be vertex `1`, however it could have label `2`. And `out_nbrs[x]` is `out_nbrs[2]`, which might have no relation to `out_nbrs[1]`, which describes the actual out-neighbours of `x`.